[DBZ-342] fix broken MySQL data type "TIME"

rk3rn3r opened the pull request:
DBZ-342: fix broken handling of MySQL data type "TIME" for negative values and positive values >= 24:00:00 hours 
@see https://issues.jboss.org/browse/DBZ-342

gunnarmorling commented:
Hey @rk3rn3r, thanks a lot for this fix! I'm going to take a closer look asap, but could you add a short description of your chosen approach (looking for a high-level description of the nature of the change)? Thanks!

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/RowDeserializers.java:
```diff
@@ -517,12 +538,56 @@ protected static int bitSlice(long value, int bitOffset, int numberOfBits, int p
+     * @return the number of nanoseconds
+     * @throws IOException if there is an error reading from the binlog event data
+     */
+    protected static int deserializeFractionalSecondsInNanosNegative(int fsp, ByteArrayInputStream inputStream) throws IOException {
```
Why is it we need to deal with nano-seconds at all? IIUC, MySQL supports only micro-second resolution for `TIME`.

rk3rn3r replied:
see my first reply.

rk3rn3r replied:
@gunnarmorling btw, the recent code was calculating nanoseconds here too.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -446,7 +451,50 @@ protected void execute() {
                                             for (int i = 0, j = 1; i != numColumns; ++i, ++j) {
-                                                row[i] = rs.getObject(j);
+                                                Column actualColumn = table.columns().get(i);
+                                                if(actualColumn.jdbcType() == Types.TIME) {
```
How about extracting all this into its own method?

gunnarmorling replied:
Would even be an idea to write separate test for that. It'd still be an IT (as DB access is needed), but separate from all the connector logic.

rk3rn3r replied:
I can do that.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -446,7 +451,50 @@ protected void execute() {
+                                                        logger.error("Could not read MySQL TIME value as UTF-8");
+                                                        throw new RuntimeException(e);
+                                                    }
+                                                    String[] time = timeString.split(":");
```
The splitting patterns should be extracted into re-used `Pattern` instances, so to avoid reparsing the regex for each execution.

rk3rn3r replied:
good point. I will do that.

Also, using KC is in comparison more a performance issue... (reading binary data from a binlog or similar, converting them into a Java type, to convert them to a byte array part on Kafka side). As stated above, that would make more sense if you can share the types in a proper implementation for Avro (and only for Avro/Avro Registry). Other serzializers don't need that type binding and deliver their own mapping anyway. Debezium types are used only as a mapping information for KC, especially when Debezium is just a source, then it's the mapping to the type instance on the Kafka side (Also why you don't have this fromLogical/toLogical implementations in the Debezium types, right?). Only Avro needs the original type here for the schema registry stuff. Would be good to decouple that, but this probably goes against the KC behaviour and confluent pushes Avro with that. Guessed.

rk3rn3r replied:
small correction, I checked the implementation of java.lang.String#split(java.lang.String, int) and it uses a much faster algorithm (iterate over the string and match the single char) when the pattern is just a single char. I will leave that implementation then, if you agree? No need for a Pattern.compile in this case, I guess?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/RowDeserializers.java:
```diff
@@ -295,16 +296,15 @@ protected static Serializable deserializeTime(ByteArrayInputStream inputStream)
+     * @return the {@link java.time.Duration} object
      * @throws IOException if there is an error reading from the binlog event data
      */
     protected static Serializable deserializeTimeV2(int meta, ByteArrayInputStream inputStream) throws IOException {
```
Having some unit tests for these methods would be great.

rk3rn3r replied:
I just added test for things that were already there. Would it make sense to create a separate ticket to add tests for all RowDeserializers? But I can still start a test suite for it, if you want to have it.

gunnarmorling replied:
Did you end up with adding a unit test for this method? As you say it'd be a good starting point (no need for adding test for all the other methods as part of this PR).

rk3rn3r commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java:
```diff
@@ -187,6 +192,11 @@ public SchemaBuilder schemaBuilder(Column column) {
         }
+        switch (column.jdbcType()) { // see https://issues.jboss.org/browse/DBZ-342
+            case Types.TIME:
+                if (adaptiveTimePrecision && column.length() <= 0) return Time.builder();
```
@gunnarmorling that's true, but I tried to find a common implementation to fulfill the handling of Durations and the base resolution of java.time.Duration is nanosecs. I can change that. But microseconds for 839 h force to use Long / INT64 anyway, so I took nanoseconds as that fits in Long/INT64 too.
I can update the PR, divide values by 1000 and use MicroTime as builder, if you want to go with microsec presicision.

gunnarmorling replied:
> I can update the PR, divide values by 1000 and use MicroTime as builder, if you want to go with microsec presicision.

Yes, that'd be my preference. We should strive to make as less behavioral changes as possible which may affect (and break) existing consumers. Transmitting nano-seconds instead of micro-seconds is such change we should avoid.

In terms of the "connect" time precision mode being obsolete, that's also my understanding. IIUC, it's there essentially to keep compatability with consumers existing before 0.2 which introduced the "adaptive" mode to avoid precision loss for `TIME(N>3)` columns. So the same motivation as above. We surely can get rid of it at some point, likely 1.0 (whenever we feel we're ready for that).

The remaining question is should we do the change to `INT64` unconditionally (again potentially breaking clients expecting `INT32` only), or should we provide an option to keep `INT32` (which of course only should be used if values don't exceed that value range). I'd be curious about @rhauch's take on that.

About that type discussion, I agree it's something to consider, but let's better have that one separate from the change at hand.

rk3rn3r replied:
@gunnarmorling @rhauch what about introducing a third time precision mode? ("connect", "adaptive" as default and making the PR compatible, and the one I suggest?) 
I think on the sink/consumer side, when not doing Avro and not knowing if the INT32/64 means seconds, micro-/milli- or nanoseconds, is really confusing. 

We would really prefer an intuitive mapping INT32 --> always seconds, INT64 --> always nanos (or millis)

What do you think?

rhauch replied:
I don't think it's really appropriate to use the Schema version, since that will almost certainly conflict with Avro schema evolution versions (if they're used).

Changing from `INT64` to `INT32` may cause schema incompatibility issues for people that are relying upon certain evolution compatibility constraints. I don't recall how Avro would view such a change -- is it backwards or forwards compatible?

Is this really another distinct time precision mode, or is it really more overriding the mapped Schema type for a given source type or (fully-qualified) column name?

rk3rn3r replied:
@gunnarmorling you mean that one?

@rhauch I would introduce a new time mode here, as this is different behaviour. Users may not realize that this is a breaking change in the interpretation of the value and values in the stream (or in Stream Processors or sinked data in a destination database) could be misinterpreted. But I will again double check this to be really sure we don't introduce more configuration complexity, if the information in the streams is not changed. Will double check the information in the stream to be 100% sure.

gunnarmorling replied:
> @gunnarmorling you mean that one?

Agh, yes :) I was looking for it three times, but somehow missed it...

gunnarmorling replied:
I agree there should be a new time/precision mode, allowing people to configure whether either to allow for the full range of values by conveying TIME columns as `int64` or to keep the current behavior, if needed.

> I think on the sink/consumer side, when not doing Avro and not knowing if the INT32/64 means seconds, micro-/milli- or nanoseconds, is really confusing.
> 
> We would really prefer an intuitive mapping INT32 --> always seconds, INT64 --> always nanos (or millis)

Ah, I'm beginning to understand where you're coming from. IMHO, a consumer shouldn't interpret the value solely on the Connect type (`int32` vs. `int64`), though. Instead, the semantical type name should be used, i.e. `io.debezium.time.Time`. The semantics of that type are that it always conveys numbers in milliseconds. And that's not going to change, no matter whether the "physical" type is `int32` or if we changed it to `int64`.

Of course this assumes the schema is present in the JSON. I don't think it's practical to have no schema info at all (how are consumers even going to know that a `int32`/`int64` column is representing a duration without any out-of-band information on the semantics of that column; for them it's just an arbitrary number).

N.b. already today, `int32` may represent days (`io.debezium.time.Date`)  or milli-seconds (`io.debezium.time.Date`). And `int64` may be milli-seconds or micro-seconds (see http://debezium.io/docs/connectors/mysql/#temporal-values). So you can't make any conclusion just from looking at the plain type without the semantical name associated to it.

But I see how it may be desirable to not change the actual type of `io.debezium.time.Time` depending on some setting. I still don't think a new type is needed, though, a we already have `io.debezium.time.MicroTime`.

So would this work:

* add a new value for `time.precision.mode`: `adaptive_compat` (better name welcome)
* the behavior for `adaptive` (the default) would change so that columns of `TIME(0-6)` will always be represented by `io.debezium.time.MicroTime` and `int64`
* the new value `adaptive_compat` would keep the current behavior (`TIME(0-3)` are represented by `io.debezium.time.Time` and `int32`)

Would that work for you, @rk3rn3r? If not, how about having a quick Google Hangout some time this week, it may be easier to converge discussing face to face :)

rhauch replied:
@gunnarmorling can we try to avoid changing behavior of existing functionality, unless doing so will result in the new Avro schema changes that are *both* backward and forward compatible with the Avro schema generated from the previous version?

I'm fine with adding a new `time.precision.mode`, but *changing* the `adaptive` is likely to generate an Avro schema that is not backward or forward compatible with the Avro schema generated for the same table by the previous Debezium version.

Changing field types is part of the API of Debezium, so we need to be far more careful about not breaking existing deployments when they upgrade.

gunnarmorling replied:
@rhauch I agree on the need on being careful here, but I don't think it's that we break things light-heartedly here (hence my proposal to have an option for keeping the current behavior).

But IMHO the current `adaptive` mode is broken, as it choses a type (`int32`) which doesn't fit the entire range of potential values for `TIME(0-3)`. So we need to provide users with a mapping of that column type which fits all the potential values, i.e. `int64`. Now the question is which setting should be the default, going forward. Based on previous similar cases, I think the correct mapping should be the default behavior, i.e. `int64`. I don't mind too much whether of the following two we do:

* having a new precision mode mapping to `int64`, using that as default and keep `adaptive` for compatability reasons, carrying the old mapping
* having `adaptive` change its mapping to `int64`, keept it as the default and have a new mode for compatability reasons, carrying the old meaning

But either way, I think `int64` should be the default if nothing else configured. Yes, it'd alter the default behavior, but it really is required to be able to convey all the possible values of that column type. In addition to the default, there'll be the option for keeping the current behavior, e.g. when upgrading an existing connector and being sure that `int32` is enough.

This change is scheduled for 0.7 so this seems the best thing to do to me, of course with appropriate documentation in the change log.

In a way it's the same kind of change as the BIGINT UNSIGNED mapping ([DBZ-363](https://issues.jboss.org/projects/DBZ/issues/DBZ-363)), where we agreed on changing the default mapping to a more sensible approach in 0.7, with an option to keep the existing behavior if required.

Does this make sense? Or do you see a difference to this case here?

rhauch replied:
It's fine, but I just wanted the decision to break compatibility to be explicit and intentional.

gunnarmorling replied:
Ok, seems we're on the same page then :) Yes, it is intentional and it will be explicit by having a prominent entry in the release notes (together with the BIGINT UNSIGNED change).

rk3rn3r commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java:
```diff
@@ -713,4 +726,62 @@ else if (data instanceof Number) {
     }
+
+    /**
+     * Converts a value object for an expected JDBC type of {@link Types#TIME} to {@link java.time.Duration} values.
```
hehe... good catch. will fix that

rk3rn3r commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/RowDeserializers.java:
```diff
@@ -295,16 +296,15 @@ protected static Serializable deserializeTime(ByteArrayInputStream inputStream)
 
     /**
-     * Converts a MySQL {@code TIME} value <em>with fractional seconds</em> to a {@link LocalTime}.
+     * Converts a MySQL {@code TIME} value <em>with fractional seconds</em> to a {@link java.time.Duration}.
```
I went with the zen of python here: Explicit is better then implicit.

rk3rn3r commented:
Sure @gunnarmorling!

My goal was to fix the issue that the current implementation was not able to handle MySQL fields of the datatype TIME/TIME(x) for positive values >= 24:00:00.000000 and negative values. MySQL TIME field allows values between -838:59:59.000000 up to 838:59.59.000000 hours (see https://dev.mysql.com/doc/refman/5.7/en/time.html ). java.sql.Time can only handle daytime (00:00:00 - 23:59:59).

I had to change Java data type from java.sql.Time to java.time.Duration as a Value Object. I had to remove using the Kafka Connect implementation (now adaptive time precision is more INT32 = seconds resolution, INT64 for nanosecs resolution vs always INT64/nanosecs), but I did not reflect this in the config yet.
I added tests that show that the recent code is broken and this PR fixes the issues.

The PR leads to a slightly different behavior in handling precision (like described above), but using INT32 for microsec precision would be broken anyway.
24h =                  86.400.000.000 microsec (old implementation where column.length <= 3 is mapped to Debezium Time type
Signed Max Int =  2.147.483.647 
This will fix things that are broken anyway and choose to do the BREAKING CHANGE.

gunnarmorling commented:
> The PR leads to a slightly different behavior in handling precision (like
described above), but using INT32 for microsec precision would be broken
anyway.

Just to make sure we're on the same page: the existing code (before your
change) doesn't use Int32 for representing microseconds. In "connect" mode
it uses Int32 to represent milliseconds (meaning there'll be a precision
loss if the TIME column has a finer resolution). In "adaptive" mode it uses
-- depending on the columns resolution -- Int64 for representing
microseconds (so that part is good), and Int32 for representing
milliseconds. That part is flawed, as Int32 isn't enough to hold all
potential milliseconds in 839 hours.

So this should be changed to Int64, too, so to be capable of storing the
full value range. Circling back to my question on making this an option, I
think we should actually do it. I.e. there'd be a third mode for
"time.precision.mode". This one would keep using Int32 for representing
milliseconds of TIME(1|2|3) columns. It shouldn't be used in general due to
the potential value overflow, but it'd be beneficial for cases where an
existing consumer expects Int32 and it's ensured that there are no values
larger than Int32 in that column (which will be the case if the TIME column
doesn't store values larger than 24h, which was the only thing we supported
until your change).

rk3rn3r commented:
@gunnarmorling hehe... I already suggested a third time precision mode in another comment. And I already changed my comment from micro to millisec. You got the update probably late. That was my mistake.

Should I do another time precision mode then?

rk3rn3r commented:
Thank you for all your feedback.
I suggest the following action points then:

- fix the c&p errors in the javadocs
- create UnitTests for TIME related functions in RowDeserializers class
- use RegexPattern in SnapshotReader
- extract value parsing in SnapshotReader and write UnitTests
- switch to use SecondDuration, MicroSecondsDuration, MilliSecondsDuration, NanoSecondsDuration
- use these new type classes to make `time.precision.mode` work as recently / expected
- add a new option to `time.precision.mode` to use the simple seconds -> INT32, nano-/microseconds -> INT64 (nanosecs can cover more then 292 years, microsecs can conver 292.277 years). I would take nanosecs

Would you agree @gunnarmorling?

gunnarmorling commented:
@rk3rn3r I think that summarizes it pretty well. On

> switch to use SecondDuration, MicroSecondsDuration, MilliSecondsDuration, NanoSecondsDuration

I don't think we transmit seconds currently. Worth double checking, but if so I'd prefer to not start doing it now.

Also I don't see the need for transmitting nano-seconds if we can't get that resolution from the DB. The current code uses it as fallback as it seems, that seems reasonable should ever something like `TIME(7)` be introduced.

> add a new option to time.precision.mode to use the simple seconds

See above on seconds. The new option value should cause all `TIME` columns to be conveyed as `Int64`. For `TIME` and `TIME(0-3)`, the transmitted value would be milliseconds, for `TIME(4-6)` the value would be microseconds. Otherwise (theoretical case for now), nano-seconds would be transmitted (as today).

This behavior should be the default, with the current `adaptive` behavior still being available for compatibility reasons. The only difference between the current "adaptive" mode and the new default mode will be the changed data type for `TIME` and `TIME(0-3)` from `Int32` to `Int64`.

Does that make sense?

rk3rn3r commented:
I updated the action point list. Do you agree now?

For conveniece with Kafka Connect and Avro it's still better to have new MilliDuration and MicroDuration types, right? Or should I reuse the Time classes, also the conversion functions inside are misleading, but need to be kept for Postgres, I guess?
What do you think @gunnarmorling ?

gunnarmorling commented:
@rk3rn3r, you've lost me a bit at "switch to use MicroDuration, MilliDuration".

What does that mean exactly? I don't think we need any new types in the `io.debezium.time` package, if that's what you mean. The only change to the schema and types would be that `io.debezium.time.Time` produces `int64` instead of `int32` in order to be able to cover the full range of milliseconds contained in 839 hours. Plus, there would be a parameter to optionally keep using `int32` in cases where one is sure that it is enough and one would prefers that for the sake of backwards-compatability.

Otherwise, the action items look good. Does this help to clear up things?

rk3rn3r commented:
@gunnarmorling I am not sure about this. Debezium is based on Kafka Connect. And confluent is pushing for Avro. This is why they combined the type serialization and deserialization in the types classes. There is always a fromLogical and a toLogical function in a type that is used to transform values from a high level data structure to a message value (normally a int32, int64, float, double, etc).
Debezium Time type represents a Unix timestamp that was converted from a java.sql.Time/Datetime. As we change that now, I guess we need to use two new classes? Especially the actual Debezium Time types are still working and can be used with other JDBC compliant databases like Postgres. 
But probably I am wrong here?

gunnarmorling commented:
@rk3rn3r I don't think a new type is needed for this. It should suffice -- based on the chosen precision mode -- to emit a different type (and version) in the schema. I.e. the builder would look like this:

    public static SchemaBuilder builder() {
        if ( precisionMode == LEGACY ) {
            return SchemaBuilder.int32()
                            .name(SCHEMA_NAME)
                            .version(1);
        }
        else {
            return SchemaBuilder.int64()
                            .name(SCHEMA_NAME)
                            .version(2);
        }
    }

At least that's my understanding of how the type system and type changes are supposed to work. @rhauch, can you confirm that?

rk3rn3r commented:
@gunnarmorling the type version is handled in a different way imho.

gunnarmorling commented:
Hey @rk3rn3r yes, I realized that, too. @rhauch recently made a comment/proposal on this, but I can't seem to find that one right now :/ I hope he can dig it out again.

rk3rn3r commented:
As the discussion up there is going in so many different directions, I want to start a bit fresh again at the end of these PR discussion.
I setup two tables to summarize the issues I actually see and that I tried to fix. The last open question is now how to solve it. As I personally found it hard to wrap my brain around this complex stuff even when I was working on it because of its complexity, I thought a table would show the impacts in a better way:

![tables](https://imgur.com/a/3X88q)
I couldn't test the behaviour of java.util.Date until now. Probably the Postgres max value of 24:00:00 will just create an overflow to the next day, and the millisec value is fine.

Some things are actually broken or incomplete, some bigger and some smaller edge cases are affected. Negative values from MySQL are broken completely. Also Kafka Connect type is affected and shouldn't be used with MySQL.

From your discussion I got, that you both are fine with a clear mentioned breaking change or implementing a new adaptive_compat mode. Did I get it right, that you prefer the breaking change and change the adaptive precision mode?
What about Postgres? My changes will only fix the MySQL behavior.
I will do the necessary changes until end of this week, when I get a feedback soon.
@gunnarmorling @rhauch

rhauch commented:
@rk3rn3r, I am fine with breaking changes in a "major" release (which before 1.0 are the 0.x.0 releases) as long as:

1. we attempt to minimize the breaking changes, preferably to those things that are incorrect/broken
2. we very clearly announce the breaking changes, how users might be impacted, and what if anything they need to do to migrate

It would be great to see a proposal that outlines this.

BTW, the Apache Kafka project's Kafka Improvement Proposal (KIP) mechanism serves exactly this purpose: present the motivation to change the public interface/behavior, the proposed approach, how this will affect end users, and the alternatives that were considered and rejected. Every KIP must be discussed and approved by vote of committers. 

I'm not suggesting that DBZ adopt that formal of a process (especially the voting part), but it'd be great to at least be a bit more formal about the proposals.

rhauch commented:
Just to be clear, @rk3rn3r has done an excellent job with this issue and PR, and has clearly documented the problem and what is broken. I definitely would like to see this fixed, and for this PR I am merely asking for a thorough description of the proposed changes and a simple analysis of how this might affect existing users. (A valid answer is "this is broken enough that DBZ likely doesn't work well with the affected types.")

Please let me know if anyone disagrees or thinks this is not necessary or that it's overkill.

gunnarmorling commented:
> I am fine with breaking changes in a "major" release (which before 1.0 are the 0.x.0 releases) as long as:
>
> 1.   we attempt to minimize the breaking changes, preferably to those things that are incorrect/broken
> 1.   we very clearly announce the breaking changes, how users might be impacted, and what if anything they need to do to migrate

+1, that's the mental model I'm working with. I think we really all agree that breaking changes should only be done after careful consideration. How about adding this to the README.md for posterity (or another place if you have good ideas)?

I don't think a very formal process is needed at this point, it should work just fine as long as we share the same principles (which I believe we do).

> @rk3rn3r has done an excellent job with this issue and PR, and has clearly documented the problem and what is broken.

Absolutely.

> for this PR I am merely asking for a thorough description of the proposed changes and a simple analysis of how this might affect existing users.

Agreed, that'd be good at this point. I think it's quite clear to me, but the discussion has been long so it'd be great to have a concise description of the planned change and the impact on consumers.

rk3rn3r commented:
I still have a question left for doing the final changes:

actual implementation:
JDBCValueConverter:
- mode adaptive, field precision <= 3 Time.builder / INT32 meaning milliseconds
--> doesn't work for MySQL because 839h in ms exceed signed INT32 ---> BROKEN
--> works for Postgres as TIME in pgsql is 0-24h, will fit inside INT32 ---> NOT BROKEN 

- mode adaptive, field precision <=6, >3 MicroTime.builder / INT64, meaning microseconds
--> works for MySQL
--> works for Pg

- mode adaptive, field precision >6, NanoTime.builder / INT64, meaning nanoseconds
--> NEVER USED?!



For solving this, I would prefer to make a breaking change here:
----------------------------------------------
- remove adaptive time precision mode/s and config completely
--> connect mode / org.apache.kafka.connect.data.Time: only works with PG (INT32 meaning milliseconds), can only work with positive values and between 0-24h, MySQL value range will exceed INT32 on milliseconds resolution and negative values aren't supported at all
--> adaptive would be the only mode left

- remove NanoTime, because it gives a double meaning on how a TIME field value represented by an INT64 is interpreted, and this value is never used

- Time.builder should stay an INT32 value, but the new meaning will be seconds (seconds after midnight of epoch), a TIME field, no matter what datasource RDBMS created a time field, when it's an INT32 it should be SECONDS precision from now on.
It will only be used for field precision / column.length() <= 0.
--> an alternative would be PG is millisecs precision, MySQL is seconds (and having a separate implementation in MySQLValueConverters class), but I think the value range should be independent from the source datastore technology

- MicroTime.builder will stay as it is and used for all field precision / column.length() > 0 always meaning microseconds resolution

This means a breaking change for PG and MySQL. Like stated before, I can also *ONLY* apply the change to MySQL in MySQLValueConverter, but then I need to add an io.debezium.time.FullSecondsTime class for MySQL INT32 values. This way can also be done with a third precision mode.

When using the Debezium JSON format, you have the relation to the original datatype and you know the meaning the INT32/INT64 value. When using Avro or Google Protocol Buffers or other formats to transport your data, you could loose the meaning. For implementing sinks to write data captured by Debezium to destination datastores or to access the data from Apache Kafka with Kafka Streams API / Stream Processors I would prefer we have a consistent meaning like INT32 always means seconds, INT64 means microseconds.

Please review my suggestion @gunnarmorling @rhauch , do you agree? Or do you want to go another way, if so, how?

gunnarmorling commented:
Hi @rk3rn3r, as dicussed above we should be very careful to reduce breaking changes as far as possible and what you suggest seems to change much more than actually needed.

>  also disable connect mode for TIME fields 

I'd prefer to not do that atm. Yes, it's broken, but that's why we have the alternative mode(s) and I don't see a strong reason to change that right now. My current thinking is that we should get rid of such legacy options when going to DBZ 1.0.

> When using Avro or Google Protocol Buffers or other formats to transport your data, you could loose that meaning.

That's not true at least for Avro and the schema registry, it will also expose the schema name to consumers. For other, custom, converters you'd have similarly to make sure to not only transmit the literal type but also the semantic type.

>  I would prefer we have a consistent meaning like INT32 always means seconds, INT64 means microseconds. 

I'm not convinced of this. It'd mean many changes as you pointed out yourself, and I don't think it'd help much. Consumers still would need some out-of-band knowledge so to know that field "foo" of type `int32` is some amount of time in seconds and not an unrelated, non-temporal numeric value. Instead, they should consider the semantic type as well to correctly interpret the values.

> remove NanoTime, because it gives a double meaning on how a TIME field value represented by an INT64 is interpreted, and this value is never used

We could indeed remove it, but then I don't see much harm in keeping it as a fallback should a DB support larger precision than `TIME(6)` at some point. As said above, I don't think we can reasonably work with the assumption that `int64` always has the same meaning.

> Time.builder should stay an INT32 value, but the new meaning will be seconds...

I don't think we should do that.

> Or do you want to go another way, if so, how?

I still think my proposal from a couple of days ago should be the way forward:

* add a new value for time.precision.mode: `adaptive_64` or `adaptive_long` or similar; this would be the default going forward
* if `adaptive_64` is used, columns of TIME(0-6) will always be represented by `io.debezium.time.MicroTime` and `int64`
* if `adaptive` is used, just as today TIME(0-3) are represented by `io.debezium.time.Time` and `int32`; and TIME(4-6) by `MicroTime` and `int64`
* if `connect` is used, its behavior is just the same as today

I.e. semantics of the existing precision modes don't get changed but a new - and let's hope finally correct one - is added. So for people using the existing modes, there's no behavioral change (including the insufficient value ranges). The only breaking change is the changed default behavior (adaptive_64 instead of adaptive), but as said I think it should be like this as the correct behavior should be the default going forward.

I understand it doesn't allow consumers to interpret values solely on the literal type, but IMHO that's just not feasible, instead the semantic type should be considered.

Could we move forward in that spirit? It seems we're discussing it for very long :) If you still have doubts, let's do a Hangout tomorrow, so to wrap up the issue eventually. Thanks!

gunnarmorling commented:
> but a new time precision mode added

> - also `connect` and `adaptive` time precision modes are broken in several use-cases, you don't want the behavior of `adaptive` mode been fixed, but a new time precision mode added (for now)?

Yes.

> - the new mode is identical to adaptive mode for DateTime/Timestamp fields and everywhere where it is used, but behavior is different for TIME?

Yes. It'd behave as the current `adaptive`, with the exception of TIME(0-3) which would not be mapped by `int32` and `i.d.t.Time` but by `int64` and `i.d.t.MicroTime`.

> - the new mode should be applied to JDBCValue converters and, by doing this, applied to MySQL *and* Postgres?

That's a good question. IIUC, we don't have the issue with the adaptive mode for Postgres, as its `TIME` column doesn't exceed 24h, hence no change would be needed. I've no strong preference either way, but I'd say let's not expose `adaptive_time_microseconds` for PG for the time being?

rk3rn3r commented:
@gunnarmorling @rhauch I recently pushed my latest state. I implemented the changes from your earlier feedback at the beginning if this PR, but some of them I dropped for several reasons. We can discuss them in detail, if you want to know why.

With the recent changes I added the `adaptive_time_microseconds` mode and made it the default mode for the MySQL Connector like we discussed it. MySQL issues are fixed now. Every connector that is based on the JDBCValueConverters can use the new mode (as long as the capturing layer doesn't mess up 24:00:00 exceeding on mapped java data types).

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/RowDeserializers.java:
```diff
@@ -322,19 +322,40 @@ protected static Serializable deserializeTimeV2(int meta, ByteArrayInputStream i
+        int seconds = bitSlice(time, 18, 6, 24);
+        int nanoSeconds;
+        if (is_negative) { // mysql binary arithmetic for negative encoded values
+            int mask10bits = (1 << 10) - 1;
```
Seems we could extract those two as constants?

rk3rn3r replied:
Will do

gunnarmorling replied:
I'll leave it to you. Having some unit tests would have been nice, but as you say we already got it covered in the ITs, so it seems not strictly needed.

rk3rn3r commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -128,6 +133,53 @@ protected void doCleanup() {
+            Blob b = rs.getBlob(fieldNo);
+            String timeString;
+            try {
+                timeString = new String(b.getBytes(1, (int) (b.length())), "UTF-8");
```
I assumed, that it's almost guaranteed to be ASCII. I think most encodings are based on ASCII. As a time field only include [0-9,:,.] that are already in ASCII, I am pretty sure I could even set ASCII without any encoding problems. What do you think?

gunnarmorling replied:
Ok.

rk3rn3r commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -128,6 +133,53 @@ protected void doCleanup() {
+                logger.error("Could not read MySQL TIME value as UTF-8");
+                throw new RuntimeException(e);
+            }
+            String[] time = timeString.split(":");
```
Are you still sure about htis? I already replied earlier, that when using a single non-regex char, the implementation of String.split doesn't use RegEx pattern and compile, but just iterates over the String, which is much faster. I don't think switching to a pattern would really speed up here?

gunnarmorling replied:
That was news to me indeed. I've still opted to use a single regex to parse the whole thing in my commit to this branch, as I felt it makes the entire logic easier to grasp down the road.

rk3rn3r commented on debezium-core/src/main/java/io/debezium/jdbc/JdbcValueConverters.java:
```diff
@@ -72,7 +72,8 @@
     protected final Logger logger = LoggerFactory.getLogger(getClass());
     private final ZoneOffset defaultOffset;
-    protected final boolean adaptiveTimePrecision;
+    protected final boolean adaptiveTimePrecisionMode;
```
I did not want to write `temporalPrecisionMode.equals(TemporalPrecisionMode.ADAPTIVE)` and `temporalPrecisionMode.equals(TemporalPrecisionMode.ADAPTIVE_TIME_MICROSECONDS)` multiple times and have it evalutated for every field that is handled by the XValueConverters.convert() and XValueConverters.schemaBuilder() calls. I think that's a much better way to do it, don't you agree?

gunnarmorling replied:
Ok.

rk3rn3r commented on debezium-core/src/main/java/io/debezium/time/Conversions.java:
```diff
@@ -92,6 +93,14 @@ protected static LocalTime toLocalTime(Object obj) {
         }
+        if (obj instanceof Duration) {
+            Long value = ((Duration) obj).toNanos();
+            if (value >= 0 && value <= 86400000000000L) {
```
just copied from the Kafka Connect Time implementation. Will do that ;-)

gunnarmorling commented on debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/BinlogReaderIT.java:
```diff
@@ -264,4 +266,92 @@ public void shouldHandleTimestampTimezones() throws Exception {
+        assertThat(c5Time.getNano()).isEqualTo(999999000);
+        assertThat(c5Time.toNanos()).isEqualTo(-3020397000001000L);
+        assertTrue(c5Time.isNegative());
+        assertThat(c5Time).isEqualTo(Duration.ofHours(-838).minusMinutes(59).minusSeconds(58).plusNanos(999999000));
```
@rk3rn3r, I've been thinking a bit more about the nano question and I think this assertion isn't correct.

Instead, it should be `minusNanos()` at the end. The reason being that a negative `TIME` value in MySQL such as `-00:00:01.500` should be interpreted as "-1.5 seconds", not as "-1 second plus 0.5 seconds" (i.e. "-0.5 seconds"). This is an example I ran in MySQL to see how it behaves:

```
select mytime, duration, addtime(mytime, duration) from customers where id  =1001;
+---------------------+---------------+---------------------------+
| mytime              | duration      | addtime(mytime, duration) |
+---------------------+---------------+---------------------------+
| 2017-11-28 12:59:40 | -00:00:01.500 | 2017-11-28 12:59:38.500   |
+---------------------+---------------+---------------------------+
```

Note how the result of the calculation is 1.5 seconds less than the original value, not -1 + 0.5 seconds.

Based on that, I think the assertions as well as the implementations should be adjusted. WDYT?

rk3rn3r replied:
@gunnarmorling you're right for the mysql part, but when I compared the inner nanoseconds value representation of the java.time.Duration (call .toNanos()), then the correct value cane only when I added not substracted the nanos. let me double-check this tomorrow.

gunnarmorling commented:
Hey @rk3rn3r, I've pushed some changes based on my remarks about the handling of second fractions to clarify what I meant. Could you do the same for the binlog reader and adjust the remaining tests accordingly? With that in place, I think it's good to merge. Thanks a lot!

gunnarmorling commented:
@rk3rn3r And it's merged :) Thanks a lot! I've pushed one commit for testing the existing adaptive mode on MySQL. While it's broken, it still should work for people using it before and who are sure that int32 is enough. Hence a test seemed justified.

The last task remaining is the docs update; could you send a PR for updating the info on temporal types in the MySQL and Postgres connector docs?
